src/blueprint.zig

```diff
@@ -0,0 +1,330 @@
+				},
+				BlueprintCompression => {
+					@field(self, field.name) = @enumFromInt(try reader.readInt(u16));
+				},
+				else => unreachable,
+			}
+		}
+	}
+	pub fn getBlockArraySize(self: @This()) usize {
+		return @as(usize, @intCast(self.blockArraySizeX))*@as(usize, @intCast(self.blockArraySizeY))*@as(usize, @intCast(self.blockArraySizeZ));
+	}
+	fn getBlockArraySizeBytes(self: @This()) usize {
+		return self.getBlockArraySize()*@sizeOf(BlockStorageType);
+	}
+	pub fn getHeaderSizeBytes(_: @This()) usize {
+		var size: usize = 0;
+		inline for(@typeInfo(@This()).@"struct".fields) |field| {
+			size += @sizeOf(field.type);
+		}
+		return size;
+	}
+	pub fn getDecompressedDataSizeBytes(self: @This()) usize {
+		return self.paletteSizeBytes + self.getBlockArraySizeBytes();
+	}
+};
+
+pub const Blueprint = struct {
+	blocks: main.List(Block),
+	sizeX: usize,
+	sizeY: usize,
+	sizeZ: usize,
+
+	pub fn init(allocator: NeverFailingAllocator) @This() {
+		return Blueprint{
+			.blocks = .init(allocator),
+			.sizeX = 0,
+			.sizeY = 0,
+			.sizeZ = 0,
+		};
+	}
+	pub fn deinit(self: *@This()) void {
+		self.blocks.deinit();
+	}
+	pub fn clear(self: *@This()) void {
+		self.sizeX = 0;
+		self.sizeY = 0;
+		self.sizeZ = 0;
+		self.blocks.clearRetainingCapacity();
+	}
+	pub fn capture(self: *@This(), pos1: Vec3i, pos2: Vec3i) void {
+		self.clear();
+
+		const startX = @min(pos1[0], pos2[0]);
+		const startY = @min(pos1[1], pos2[1]);
+		const startZ = @min(pos1[2], pos2[2]);
+
+		const endX = @max(pos1[0], pos2[0]);
+		const endY = @max(pos1[1], pos2[1]);
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: usize = @intCast(@abs(endX - startX + 1));
+		const sizeY: usize = @intCast(@abs(endY - startY + 1));
+		const sizeZ: usize = @intCast(@abs(endZ - startZ + 1));
+
+		self.sizeX = sizeX;
+		self.sizeY = sizeY;
+		self.sizeZ = sizeZ;
+
+		for(0..sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = main.server.world.?.getBlock(worldX, worldY, worldZ) orelse Block{.typ = 0, .data = 0};
+					self.blocks.append(block);
+				}
+			}
+		}
+	}
+	pub fn paste(self: @This(), pos: Vec3i) void {
+		const startX = pos[0];
+		const startY = pos[1];
+		const startZ = pos[2];
+
+		const sizeX: usize = self.sizeX;
+		const sizeY: usize = self.sizeY;
+		const sizeZ: usize = self.sizeZ;
+
+		var blockIndex: usize = 0;
+
+		for(0..sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = self.blocks.items[blockIndex];
+					mesh_storage.updateBlock(worldX, worldY, worldZ, block);
```

Comment:
This will fail on headless servers.

Reply:
How do I check if server is headless? Or just how do I check if this is allowed?

Reply:
This is never allowed. You should never try to access client-side data structures from server side code.
In fact in this case this would even lead to updating the data twice, once through network interface, and once here.

Reply:
So should I just discard this line?

Reply:
yes

Reply:
```suggestion
```

By the way, just so you know, you can create directly committable code change suggestions in Github UI. You have to create code block with `suggestion` as language or use a button in UI, then It also copies current code automatically. You can select multiple lines too. I like this a lot when I'm doing reviews.

![image](https://github.com/user-attachments/assets/afe79971-e643-482a-bb8d-6947a22fca7d)

Reply:
Oh I didn't know that. I'll try that out.